lib/travis/model/repository/settings.rb

```diff
@@ -38,13 +39,32 @@ def validate_correctness
       # it seems there is no easy way to check if key
       # needs a pass phrase with ruby's openssl bindings,
       # that's why we need to manually check that
-      if value.decrypt.to_s =~ /ENCRYPTED/
-        errors.add(:value, :key_with_a_passphrase)
-      else
-        errors.add(:value, :not_a_private_key)
+      unless valid_nonrsa?
+       if value.decrypt.to_s =~ /ENCRYPTED/
+          errors.add(:value, :key_with_a_passphrase)
+        else
+          errors.add(:value, :not_a_private_key)
+        end
       end
     end
 
+    def valid_nonrsa?
+      key = SSHData::PrivateKey.parse_openssh(value.decrypt)
+      unless key
+        if value.decrypt.to_s =~ /ENCRYPTED/
+          errors.add(:value, :key_with_a_passphrase)
+        else
+          errors.add(:value, :not_a_private_key)
+        end
+        return false
+      end
+
+      true
+    rescue SSHData::DecodeError
+      false
+    end
+
+
```

Comment:
```ruby
def validate_nonrsa
  key = SSHData::PrivateKey.parse_openssh(value.decrypt)
  unless key
    add_errors
  end
rescue SSHData::DecodeError
  add_errors
end

def add_errors
  if value.decrypt.to_s =~ /ENCRYPTED/
    errors.add(:value, :key_with_a_passphrase)
  else
    errors.add(:value, :not_a_private_key)
  end
end
```
and replace this `unless valid_nonrsa?`
with
`validate_nonrsa`

What do you think?